src/apify_client/_http_client.py

```diff
@@ -0,0 +1,108 @@
+import re
+from typing import Any, Callable, Dict, List, Optional, Union
+
+import requests
+from requests.exceptions import ConnectionError, Timeout
+
+from ._errors import ApifyApiError, InvalidResponseBodyError
+from ._utils import _retry_with_exp_backoff
+
+NO_CONTENT_STATUS_CODE = 204
+RATE_LIMIT_EXCEEDED_STATUS_CODE = 429
+
+
+class _HTTPClient:
+    def __init__(self, max_retries: int = 8, min_delay_between_retries_millis: int = 500) -> None:
+        self.max_retries = max_retries
+        self.min_delay_between_retries_millis = min_delay_between_retries_millis
+        self.requests_session = requests.Session()
+
+    def call(
+        self,
+        *,
+        method: str,
+        url: str,
+        headers: Dict = None,
+        params: Dict = None,
+        data: str = None,
+        json: Union[str, Dict, List[str], List[Dict]] = None,
+        stream: bool = None,
+        parse_response: bool = True,
+    ) -> requests.models.Response:
+        request_params = self._parse_params(params)
+        requests_session = self.requests_session
+
+        def _make_request(bail: Callable, attempt: int) -> requests.models.Response:  # type: ignore[return]
```

Comment:
We are currently forcing any retryable function to have `bail` and `attempt` args an no other args. I don't think this is very flexible.